### PR TITLE
Big mock-server testing refactor...

### PR DESCRIPTION
**Pull request by johnstonematt**
# Note

Just throwing this in here - apologies in advance, I know this PR is big and should've been broken into many smaller ones,  so I'm sorry for that. I figured the whole thing out in a rather haphazard manner, hence why it all sort've got lumped together in 1 big boy.

That being said, the changes here are almost exclusively to the tests. The changes to non-test code are almost all just removals of redundant code to make writing tests easier.

# Goal

The main goal of this beefy PR was to introduce a server that can mock an RPC node, and to write the tests off that. The bonuses of this are:
* Dries up the code, as the provider interface is now no longer needed.
* Makes future tests easier to write.

# Changes
#### Client

Changes here were generally focused around:
* Simplifying the method signatures by:
  * Removing arguments that were never used.
  * Switching arguments that were pointers for the sake of being nullable back to "regular" types.
* Removing the `Provider` interface.

#### Server

The main change here is the addition of `mock.go`, which contains a local server that mocks an RPC node. The `MockServer` contains logic for simulating the dynamically changing responses of an RPC node, as opposed to just static pre-defined request handling.

### Tests

Tests were added that check the internal method logic and data parsing of the `Client`, using the `MockServer`. 

### Main Package

Changes here mostly revolved around implementing the changes to the `rpc` package. But included some other minor ones, such as:
* Standardising on `nodekey` instead of `identity` (to match the metric labels)
* Switching the `BlockSizeMetric` labels from `vote/total` to `vote/non_vote`
* Moving `main()` to `main.go`

#### Slots & Exporter

This is where the big changes happened. I would advise reading these files "as is" instead of the diff. 

Besides the `MockServer` refactor, the following changes were made to the tests:
* `TestSolanaCollector_Static` was removed as this is a very simple variation of `TestSolanaCollector_Dynamic` (the collection in the tests happens on demand)
* Tests on the newer metrics were added.

#### Utils

The main change here is the addition of `TestCountVoteAccounts`, which tests the function on a real block. The expected result is from the explorer.

#### Config

Added `config_test.go` which tests `NewExporterConfig`. This is a very simple function being tested and so I don't expect this to ever fail, but I figured better to have this than not.

### CODE REVIEW

**Comment by johnstonematt** [comment]
I accidentally merged the README update into here - probably should've left it separate (this is bloated enough as is), but just know you can ignore that.

File: cmd/solana_exporter/utils_test.go [hunk 1]
```diff
@@ -63,3 +88,17 @@ func TestGetEpochBounds(t *testing.T) {
 	assert.Equal(t, int64(20), first)
 	assert.Equal(t, int64(29), last)
 }
+
+//go:embed testdata/block-297609329.json
```
Comment (SEJeff) [inline cmd/solana_exporter/utils_test.go:92]: Nice use of go language features.

File: cmd/solana_exporter/utils_test.go [hunk 2]
```diff
@@ -63,3 +88,17 @@ func TestGetEpochBounds(t *testing.T) {
...
+func TestCountVoteTransactions(t *testing.T) {
+	var block rpc.Block
+	err := json.Unmarshal(blockJson, &block)
+	assert.NoError(t, err)
+
+	voteCount, err := CountVoteTransactions(&block)
+	assert.NoError(t, err)
+	// https://explorer.solana.com/block/297609329
+	assert.Equal(t, 1048, voteCount)
+}
```
Comment (SEJeff) [inline cmd/solana_exporter/utils_test.go:104]: Can you also test the non-vote transactions via a length check like you did here please?

File: pkg/rpc/client_test.go [hunk 3]
```diff
@@ -0,0 +1,245 @@
+package rpc
+
+import (
+	"context"
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+func newMethodTester(t *testing.T, method string, result any) (*MockServer, *Client) {
+	return NewMockClient(t, map[string]any{method: result}, nil, nil, nil, nil)
```
Comment (SEJeff) [inline pkg/rpc/client_test.go:10]:
```suggestion
	t.Helper()
	return NewMockClient(t, map[string]any{method: result}, nil, nil, nil, nil)
```

File: pkg/rpc/mock.go
Comment (SEJeff) [inline pkg/rpc/mock.go:1]: I'd almost be inclined to make this a separate package named rpcmock or something (this is how this is often done upstream, e.g.: net/http vs httptest)

**Review by SEJeff — APPROVED** [review]
Heroic effort here. Nice job, though I'd still prefer something like httptest this is a solid approach.

File: cmd/solana_exporter/testdata/block-297609329.json
[the remark this answers is not in the document]
Comment (johnstonematt) [reply cmd/solana_exporter/testdata/block-297609329.json:1]: I also don't love it, but wasn't sure how else to go about it. If you think it should be removed or replaced with some other test, am super happy to. I just thought it a fairly important test.

File: cmd/solana_exporter/utils_test.go (hunk 2 above)
Comment (johnstonematt) [reply cmd/solana_exporter/utils_test.go:104]: sure can do

**Pull request merged**